Changed the "fullscreen" config variable to an enum

Berulacks opened the pull request:
...Which can be either "Fullscreen", "FullscreenDesktop", or "Windowed".
I also udated the win32gl adapter to reflect this.
Also cleaned up some cruft in sdl.d and changed the two deprecated
logFatal() statements in linux.d to fatal()s.

Berulacks committed:
Changed the "fullscreen" config variable to an enum

...Which can be either "Fullscreen", "FullscreenDesktop", or "Windowed".
I also udated the win32gl adapter to reflect this.
Also cleaned up some cruft in sdl.d and changed the two deprecated
logFatal() statements in linux.d to fatal()s.

PxlBuzzard commented:
LGTM :+1:

dislogical commented:
The difference between Fullscreen and FullscreenDesktop isn't immediately apparent. Could we do Fullscreen and FullscreenWindowed instead? That seems to be the accepted standard in games at this point.

dislogical commented on source/dash/graphics/adapters/adapter.d:
```diff
@@ -16,7 +16,8 @@ abstract class Adapter
     uint _height, _screenHeight;
-    bool _fullscreen, _backfaceCulling, _vsync;
+    bool _backfaceCulling, _vsync;
+    WindowType _mainWindowType;
```
Why `_mainWindowType`? Why not just `_windowType`?

dislogical commented on source/dash/graphics/adapters/sdl.d:
```diff
@@ -103,7 +92,24 @@ public:
+                break;
+            case WindowType.FullscreenDesktop:
+                trace("Setting main window to FullscreenDesktop mode!");
+                window.setFullscreenSetting( SDL_WINDOW_FULLSCREEN_DESKTOP );
```
Based on [the SDL docs](https://wiki.libsdl.org/SDL_WindowFlags), it seems like all this does is just use the default desktop resolution. It seems to me that we want `SDL_WINDOW_BORDERLESS` at the native resolution instead. That is, assuming we rename this to `FullscreenWindowed`.

dislogical commented on source/dash/graphics/adapters/win32gl.d:
```diff
@@ -225,7 +225,7 @@ public:
         }
 
-        if( _wasFullscreen != fullscreen )
+        if( _wasFullscreen != ( mainWindowType == WindowType.Fullscreen) )
```
Spacing here is weird? Looks like you're missing a space after `Fullscreen`.

dislogical commented on source/dash/graphics/adapters/win32gl.d:
```diff
@@ -234,7 +234,7 @@ public:
         }
 
-        _wasFullscreen = fullscreen;
+        _wasFullscreen = ( mainWindowType == WindowType.Fullscreen);
```
Ditto.

dislogical commented on source/dash/graphics/adapters/win32gl.d:
```diff
@@ -281,7 +281,7 @@ public:
     final void openWindow( bool showWindow )
     {
-        hWnd = CreateWindowEx( 0, DGame.instance.title.ptr, DGame.instance.title.ptr, fullscreen ? DWS_FULLSCREEN : DWS_WINDOWED,
+        hWnd = CreateWindowEx( 0, DGame.instance.title.ptr, DGame.instance.title.ptr, (mainWindowType == WindowType.Fullscreen ) ? DWS_FULLSCREEN : DWS_WINDOWED,
```
Missing a space before `mainWindowType`.

dislogical commented on source/dash/utility/config.d:
```diff
@@ -17,6 +17,15 @@ const(Config) config() @property
     return configInst;
 }
 
+
```
Too many blank lines here.

dislogical commented on source/dash/utility/config.d:
```diff
@@ -61,12 +70,13 @@ public:
         uint height = 1920;
         @rename( "Width" ) @optional
         uint width = 720;
+
```
Shouldn't be a blank line here.

Berulacks commented:
> Based on the SDL docs, it seems like all this does is just use the default desktop resolution. It seems to me that we want SDL_WINDOW_BORDERLESS at the native resolution instead. That is, assuming we rename this to FullscreenWindowed.

Aha. Good point. I'll add a Borderless Window option.

> Why _mainWindowType? Why not just _windowType?

A variable called _windowType of type WindowType felt a bit off to me. Just semantics, really. If you prefer _windowType I'll switch it to that.

Berulacks commented:
> Shouldn't be a blank line here.

I regret everything I've ever done.

dislogical commented:
I would prefer `_windowType`, "main" doesn't mean or add anything, and just makes it longer.

Just so you know, you can directly reply to the line notes.

Berulacks commented:
> Just so you know, you can directly reply to the line notes.

:+1:  Cheers, totally missed that.

Berulacks committed:
Added borderless window option, cleaned up code and naming conventions.

Removed an unnecessary variable.
Removed FullscreenDesktop mode, replaced it with FullscreenWindowed.
Updated call to gfm's "window.setBordered"

Berulacks commented:
Should be good, now.
http://youtu.be/GsjvXnjkTLA

PxlBuzzard commented:
@Berulacks You uploaded that video just for this comment? That's some commitment.

Berulacks commented:
No one on the Internet had uploaded that clip, yet (with relatively decent quality). I was feeling productive.

dislogical commented:
I still don't think we need `FullscreenDesktop`. Seems like we should just merge it with `BorderlessWindow` to make `FullscreenWindowed`, that's just a borderless window, but add native resolution.

Berulacks commented:
## Do we really want that, though? I mean, think about it: the full screen desktop "feature" exists, I'm assuming, as a convenience, in SDL. If that's the case, then wouldn't full-screen windowed fall under the same category of a minor convenience? I feel like it might be best to either include both options, or neither. Though, I'm probably thinking too much into this. I'll make a commit when my plane lands. 

-Beru 
Sent from my mobile phone, please excuse my brevity.

On December 18, 2014 5:57:09 PM EST, Colden Cullen notifications@github.com wrote:

> I still don't think we need `FullscreenDesktop`. Seems like we should
> just merge it with `BorderlessWindow` to make `FullscreenWindowed`,
> that's just a borderless window, but add native resolution.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/Circular-Studios/Dash/pull/238#issuecomment-67571955

dislogical commented:
I mean, isn't `FULLSCREEN_DESKTOP` just the same as `FULLSCREEN` at the window resolution? If so, it's probably not worth adding a whole option for it, especially one we have to maintain. My vote: Remove `FullscreenDesktop`, rename `BorderlessWindowed` to `FullscreenWindowed`, and make it adopt the screen resolution regardless of config settings.

patrickjm committed:
Fixed documentation link in README

dislogical committed:
[CI] Re-enabled OSX build

Berulacks commented:
Wow. Totally forgot about this, my bad. I'm removing FullscreenDesktop and creating a FullscreenWindowed. Are we sure we don't want to keep a borderless windowed option, though? Seems like something some people might want it.

dislogical commented:
I do like the idea of BorderlessWindow, it was mostly the FullscreenDesktop thing I thought seemed redundant.

Nice work, btw. Currently in the Chicago airport, but I should be able to do a final-ish review on Tuesday or Wednesday.

Berulacks commented:
:+1: :+1:

dislogical commented on source/dash/graphics/adapters/sdl.d:
```diff
@@ -103,7 +92,35 @@ public:
+                bordered = false;
+                width = screenWidth;
+                height = screenHeight;
+                goto default;
```
I like that this is a reasonable use of `goto`.

dislogical commented on source/dash/graphics/adapters/sdl.d:
```diff
@@ -103,7 +92,35 @@ public:
+            default: 
+                window.setFullscreenSetting( 0 );
+                window.setSize( width, height );
+                SDL_SetWindowBordered( SDL_GetWindowFromID( window.id() ), bordered );
```
This looks pretty grody. Is there any way to get a `window.setWindowBordered` (or similar) into `gfm`?

dislogical commented:
Aside from the `setWindowBordered` thing, LTGM :+1:

Berulacks commented:
## I actually already sent them a pull request, and it was accepted. I'm heading home right now, when I get there I'll update the gfm dependency and change that. 

-Beru 
Sent from my mobile phone, please excuse my brevity.

On January 14, 2015 9:26:09 PM EET, Colden Cullen notifications@github.com wrote:

> Aside from the `setWindowBordered` thing, LTGM :+1: 
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/Circular-Studios/Dash/pull/238#issuecomment-69975181

dislogical commented:
Looks good! Also looks like you've got conflicts... Could you rebase, and we'll merge? While you're at it, if you wanted to squash a few of those commits it would be highly appreciated.

Berulacks commented:
Squashed a few commits, how's it look now? I can squash any others, if need be. Sorry for taking so long... this whole squashing business was rather new to me :)

dislogical commented:
I'm confused... there are 12 commits now? Not sure how that happened. I'll see if I can clean it up.

Berulacks commented:
## Haha. Man, I'm bad at this rebasing business. I'll try and figure out what I did wrong when I get out of this here class. I guess when I squashed the commits git duplicated them, somehow? My mistake. 

-Beru 
Sent from my mobile phone, please excuse my brevity.

On January 28, 2015 12:57:40 PM EST, Colden Cullen notifications@github.com wrote:

> I'm confused... there are 12 commits now? Not sure how that happened.
> I'll see if I can clean it up.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/Circular-Studios/Dash/pull/238#issuecomment-71882487

dislogical commented:
Alright, no worries. Ping us when it looks how you want it to look. Also worth noting you may have to do a force push for GitHub to take it.

Berulacks committed:
[Adapters] Upgraded to newest version of SDL

Conflicts:
	dub.json

Berulacks commented:
Rebasing/squashing, it seems, makes it so that the latest two commits by you guys seem to get added to the PR? It's odd. Though, the duplication happened because I did a "git pull" after rebasing, then pushing. How does it look, now? Do you think the fact that a55726e and 62ee2a4 are part of the PR would cause issues?

dislogical commented:
I think it would be ok, but I think that that's probably a sign that something else isn't right. Are you rebasing on top of `develop`?

Berulacks commented:
I was rebasing on top of feature/window_type

dislogical commented:
Oooh, you should be rebasing feature/window_type on top of master. At this point, let's just merge it.

Berulacks commented:
Man, I'm such a scrub. Oh well. Back to the git manual.
